src/query/src/dist_plan/predicate_extractor.rs

```diff
@@ -0,0 +1,1199 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Predicate extraction for partition pruning.
+//!
+//! [`PredicateExtractor`] extract a list of [`PartitionExpr`] from given [`LogicalPlan`].
+
+use std::collections::HashSet;
+
+use arrow::datatypes::DataType;
+use common_telemetry::debug;
+use datafusion_common::Result as DfResult;
+use datafusion_expr::{Expr, LogicalPlan, Operator};
+use datatypes::value::Value;
+use partition::expr::{Operand, PartitionExpr, RestrictedOp};
+
+/// Extracts a list of [`PartitionExpr`] from given [`LogicalPlan`]
+pub struct PredicateExtractor;
+
+impl PredicateExtractor {
+    /// Extract partition expressions for partition columns from logical plan  
+    /// This method returns PartitionExpr objects suitable for ConstraintPruner
+    pub fn extract_partition_expressions(
+        plan: &LogicalPlan,
+        partition_columns: &[String],
+    ) -> DfResult<Vec<PartitionExpr>> {
+        // Collect all filter expressions from the logical plan
+        let mut filter_exprs = Vec::new();
+        Self::collect_filter_expressions(plan, &mut filter_exprs)?;
+
+        if filter_exprs.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Convert each DataFusion filter expression to PartitionExpr
+        let mut partition_exprs = Vec::with_capacity(filter_exprs.len());
+        let partition_set: HashSet<String> = partition_columns.iter().cloned().collect();
+
+        for filter_expr in filter_exprs {
+            match DataFusionExprConverter::convert(&filter_expr) {
+                Ok(partition_expr) => {
+                    // Check expression for safe partition pruning
+                    match ExpressionChecker::check_expression_for_pruning(
+                        &partition_expr,
+                        &partition_set,
+                    ) {
+                        ExpressionCheckResult::UseAsIs(expr) => {
+                            partition_exprs.push(expr);
+                        }
+                        ExpressionCheckResult::UsePartial(exprs) => {
+                            partition_exprs.extend(exprs);
+                        }
+                        ExpressionCheckResult::Drop => {
+                            debug!(
+                                "Dropping mixed expression for correctness: {}",
+                                partition_expr
+                            );
+                        }
+                    }
+                }
+                Err(err) => {
```

Comment:
what will happen if a complex filter expression involve partition columns but can't be converted to restricted partition exprs?

Reply:
it will be ignored as this expr doesn't exist

Reply:
> it will be ignored as this expr doesn't exist

is it ok to ignore it though? If the filter is something like `a>=30 or black_box(a)` would it be handled correctly?